provenance/core.py

```diff
@@ -91,23 +96,29 @@ def fn_info(f):
     return info
 
 
-def hash_inputs(inputs):
+def hash_inputs(inputs, check_mutations=False):
     kargs = {}
     varargs = []
-    all_ids = frozenset()
+    all_artifacts = {}
 
     for k, v in inputs['kargs'].items():
-        h, ids = hash(v, hasher=ah.artifact_id_hasher())
+        h, artifacts = hash(v, hasher=ah.artifact_hasher())
         kargs[k] = h
-        all_ids |= ids
+        all_artifacts = t.merge(all_artifacts, artifacts)
 
     for v in inputs['varargs']:
-        h, ids = hash(v, hasher=ah.artifact_id_hasher())
+        h, artifacts = hash(v, hasher=ah.artifact_hasher())
         varargs.append(h)
-        all_ids |= ids
+        all_artifacts = t.merge(all_artifacts, artifacts)
+
+    if check_mutations:
+        for a in all_artifacts.values():
+            if a.value_id != hash(a.value):
+                msg = "Artifact {}, of type {} has been mutated"
```

Comment:
In this loop we no longer know which argument the artifact came from. `all_artifacts` is a map of `artifact_id -> artifact` which is serving as a poor mans set because python doesn't allow us to use a set for objects. This is why the artifact_hasher api was changed to return a map instead of a set when access to the artifacts was required instead of just access to the artifact ids.

I probably shouldn't have let the map leak from outside of the artifact_hasher api though.

We could still mirror the ImpureFunctionError message, but it would require tracking which argument each artifact came from, with each artifact possibly having possibly come from multiple artifacts.